backends/ebpf/psa/externs/ebpfPsaMeter.cpp

```diff
@@ -0,0 +1,435 @@
+                           CodeGenInspector *codeGen)
+        : EBPFTableBase(program, instanceName, codeGen) {
+    CHECK_NULL(di);
+    auto typeName = di->type->toString();
+    if (typeName == "DirectMeter") {
+        isDirect = true;
+    } else if (typeName.startsWith("Meter")) {
+        isDirect = false;
+        auto ts = di->type->to<IR::Type_Specialized>();
+        this->keyArg = ts->arguments->at(0);
+        this->keyType = EBPFTypeFactory::instance->create(keyArg);
+
+        auto declaredSize = di->arguments->at(0)->expression->to<IR::Constant>();
+        if (!declaredSize->fitsUint()) {
+            ::error(ErrorType::ERR_OVERLIMIT, "%1%: size too large", declaredSize);
+            return;
+        }
+        size = declaredSize->asUnsigned();
+    } else {
+        ::error(ErrorType::ERR_INVALID, "Not known Meter type: %1%", di);
+        return;
+    }
+
+    auto typeExpr = di->arguments->at(isDirect ? 0 : 1)->expression->to<IR::Constant>();
+    this->type = toType(typeExpr->asInt());
+}
+
+EBPFType * EBPFMeterPSA::getBaseValueType(P4::ReferenceMap* refMap) {
+    IR::IndexedVector<IR::StructField> vec = getValueFields();
+    auto valueStructType = new IR::Type_Struct(
+            IR::ID(getBaseStructName(refMap)), vec);
+    return EBPFTypeFactory::instance->create(valueStructType);
+}
+
+EBPFType * EBPFMeterPSA::getIndirectValueType() const {
+    auto vec = IR::IndexedVector<IR::StructField>();
+
+    auto baseValue = new IR::Type_Struct(IR::ID(getBaseStructName(program->refMap)));
+    vec.push_back(new IR::StructField(IR::ID(indirectValueField), baseValue));
+
+    IR::Type_Struct *spinLock = createSpinlockStruct();
+    vec.push_back(new IR::StructField(IR::ID(spinlockField), spinLock));
+
+    auto valueType = new IR::Type_Struct(
+            IR::ID(getIndirectStructName()), vec);
+    auto meterType = EBPFTypeFactory::instance->create(valueType);
+
+    return meterType;
+}
+
+cstring EBPFMeterPSA::getBaseStructName(P4::ReferenceMap* refMap) {
+    static cstring valueBaseStructName;
+
+    if (valueBaseStructName.isNullOrEmpty()) {
+        valueBaseStructName = refMap->newName("meter_value");
+    }
+
+    return valueBaseStructName;
+}
+
+cstring EBPFMeterPSA::getIndirectStructName() const {
+    static cstring valueIndirectStructName;
+
+    if (valueIndirectStructName.isNullOrEmpty()) {
+        valueIndirectStructName = program->refMap->newName("indirect_meter");
+    }
+
+    return valueIndirectStructName;
+}
+
+EBPFMeterPSA::MeterType EBPFMeterPSA::toType(const int typeCode) {
+    if (typeCode == 0) {
+        return PACKETS;
+    } else if (typeCode == 1) {
+        return BYTES;
+    } else {
+        BUG("Unknown meter type %1%", typeCode);
+    }
+}
+
+IR::IndexedVector<IR::StructField> EBPFMeterPSA::getValueFields() {
+    auto vec = IR::IndexedVector<IR::StructField>();
+    auto bits_64 = new IR::Type_Bits(64, false);
+    vec.push_back(new IR::StructField(IR::ID("pir_period"), bits_64));
```

Comment:
This would be more maintainable if you put these fields into a list initializer with strings and used a loop to create vec. All the types seem to be bits_64.